src/main.rs

```diff
@@ -1,3 +1,204 @@
+        Diagnostic, DiagnosticSeverity, DidChangeTextDocumentParams, DidOpenTextDocumentParams,
+        InitializeParams, InitializeResult, InitializedParams, NumberOrString, Position, Range,
+        ServerCapabilities, TextDocumentSyncCapability, TextDocumentSyncKind, Url,
+    },
+    Client, LanguageServer, LspService, Server,
+};
+use tree_sitter_lint::{
+    tree_sitter::{self, InputEdit, Parser, Point, Tree},
+    tree_sitter_grep::{Parseable, SupportedLanguage},
+};
+
+#[derive(Debug)]
+struct Backend {
+    client: Client,
+    per_file: DashMap<Url, PerFileState>,
+}
+
+impl Backend {
+    pub fn new(client: Client) -> Self {
+        Self {
+            client,
+            per_file: Default::default(),
+        }
+    }
+
+    async fn run_linting_and_report_diagnostics(&self, uri: &Url) {
+        let per_file_state = self.per_file.get(uri).unwrap();
+        let violations = tree_sitter_lint_local::run_for_slice(
+            &per_file_state.contents,
+            Some(&per_file_state.tree),
+            "dummy_path",
+            Default::default(),
+        );
+        self.client
+            .publish_diagnostics(
+                uri.clone(),
+                violations
+                    .into_iter()
+                    .map(|violation| Diagnostic {
+                        message: violation.message,
+                        range: tree_sitter_range_to_lsp_range(
+                            &per_file_state.contents,
+                            violation.range,
+                        ),
+                        severity: Some(DiagnosticSeverity::ERROR),
+                        code: Some(NumberOrString::String(violation.rule.name)),
+                        source: Some("tree-sitter-lint".to_owned()),
+                        ..Default::default()
+                    })
+                    .collect(),
+                None,
+            )
+            .await;
+    }
+}
+
+#[tower_lsp::async_trait]
+impl LanguageServer for Backend {
+    async fn initialize(&self, _: InitializeParams) -> Result<InitializeResult> {
+        Ok(InitializeResult {
+            capabilities: ServerCapabilities {
+                text_document_sync: Some(TextDocumentSyncCapability::Kind(
+                    TextDocumentSyncKind::INCREMENTAL,
+                )),
+                ..Default::default()
+            },
+            ..Default::default()
+        })
+    }
+
+    async fn initialized(&self, _: InitializedParams) {
+        // self.client
+        //     .log_message(tower_lsp::lsp_types::MessageType::INFO, "server initialized!")
+        //     .await;
+    }
+
+    async fn shutdown(&self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn did_open(&self, params: DidOpenTextDocumentParams) {
+        let contents: Rope = (&*params.text_document.text).into();
+        self.per_file.insert(
+            params.text_document.uri.clone(),
+            PerFileState {
+                tree: parse_from_scratch(&contents),
```

Comment:
I don't know if this wrapper method is necessary vs just `parse(&contents, None)`